### PR TITLE
Improve usage of fenced divs for Theorem and Proof environment 

### LINKED ISSUE #1143: don't respect beamer theme's buildin theorem/proof block
here is a minimal, self-contained, and reproducible example

````
---
title: "beamer"
documentclass: beamer
output: 
  bookdown::pdf_book: 
    base_format: rmarkdown::beamer_presentation
    toc: no
    latex_engine: xelatex
    citation_package: natbib
    theme: Xiaoshan # https://www.ctan.org/pkg/pgfornament-han 
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from beamer theme [*Xiaoshan*](https://github.com/liantze/pgfornament-han/)

::: {.block data-latex="{Metropolis}"}
block is ok.
:::


::: {.exampleblock data-latex="{Metropolis}"}
exampleblock is also ok.
:::

::: {.theorem data-latex="{Metropolis}"}
theorem is not ok.
:::

````

my session info

```r
xfun::session_info('bookdown')
```
```
R version 4.0.4 (2021-02-15)
Platform: x86_64-apple-darwin20.3.0 (64-bit)
Running under: macOS Big Sur 10.16, RStudio 1.4.1106

Locale: en_US.UTF-8 / en_US.UTF-8 / en_US.UTF-8 / C / en_US.UTF-8 / en_US.UTF-8

Package version:
  base64enc_0.1.3   bookdown_0.21     digest_0.6.27     evaluate_0.14     glue_1.4.2       
  graphics_4.0.4    grDevices_4.0.4   highr_0.9         htmltools_0.5.1.1 jsonlite_1.7.2   
  knitr_1.32        magrittr_2.0.1    markdown_1.1      methods_4.0.4     mime_0.10        
  rlang_0.4.10      rmarkdown_2.7     stats_4.0.4       stringi_1.5.4     stringr_1.4.0    
  tinytex_0.31      tools_4.0.4       utils_4.0.4       xfun_0.22         yaml_2.2.1      
```

This bug also reported [here](https://d.cosx.org/d/421591-beamer/6)

### PR DESCRIPTION
**Pull request by cderv**
This PR closes #1143 

After all I think it is best the **bookdown** does not interfere with _beamer_. _beamer_ defines its [own theorem env](https://github.com/josephwright/beamer/blob/56b5d77a5b1ad886f452f23148b0e2e36094c689/base/beamerbasetheorems.sty#L114) with some not supported by **bookdown**. So it would be lead to have a specific syntax for some env supported by **bookdown** and another syntax for other 

````markdown
::: {.lemma name="my name"}
`lemma` is known by bookdown
:::

::: {.fact data-latex='[Special]'}
fact is not known but define by beamer. It would only work in beamer.
:::
```` 

If we don't process beamer output, this means that the syntax is 

````markdown
::: {.lemma data-latex="[my name]"}
Using `data-latex` works also as `latex`
:::

::: {.fact data-latex='[Special]'}
Using beamer own definintion environment not in **bookdown** works with the same syntax
:::
```` 

The syntax offer by **bookdown** is there because we want the same type of output for these supported environment for HTML and PDF book. I am not sure that when **bookdown** is used with _beamer_, the `gitbook()` format is also one output. is it ? 

The Lua filter is now easy to change - providing support for beamer with bookdown syntax would mean changing this only 

````diff
diff --git a/inst/rmarkdown/lua/custom-environment.lua b/inst/rmarkdown/lua/custom-environment.lua
index d7d389e5..4986f9d2 100644
--- a/inst/rmarkdown/lua/custom-environment.lua
+++ b/inst/rmarkdown/lua/custom-environment.lua
@@ -133,7 +133,7 @@ Div = function (div)
     print_debug("label for reference -> ", label)

     -- TODO: should we support beamer also ?
-    if (FORMAT:match 'latex') then
+    if (FORMAT:match 'latex' or FORMAT:match 'beamer') then
         local label_part
         if label then
             label_part = string.format("\\protect\\hypertarget{%s}{}\\label{%s}", label, label)
@@ -192,7 +192,7 @@ Div = function (div)
 end

 -- only run filter for supported format
-if (FORMAT:match 'html' or FORMAT:match 'latex') then
+if (FORMAT:match 'html' or FORMAT:match 'latex' or FORMAT:match 'beamer') then
     return {{Meta = Meta}, {Div = Div}}
 else
     print_debug("Lua Filter skipped. Output format not supported:", FORMAT)
````

So @yihui this is as we discussed : skip **bookdown** Lua custom environment processing for _beamer_ usage. 
I think you know the pro & cons now.

@XiangyunHuang, the below is what is working now with this PR what are your thought ? 

## Examples

````markdown
---
title: "beamer"
output: 
  bookdown::beamer_presentation2:
    toc: no
    latex_engine: xelatex
    citation_package: natbib
    theme: Xiaoshan # https://www.ctan.org/pkg/pgfornament-han 
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from beamer theme [*Xiaoshan*](https://github.com/liantze/pgfornament-han/)

::: {.block latex="{Metropolis}"}
block is ok.
:::

::: {.exampleblock latex="{Metropolis}"}
exampleblock is also ok.
:::

::: {.theorem latex="[Metropolis]"}
theorem is now ok.
:::

## Some other blocks

::: {.lemma data-latex="[Metropolis]"}
Using `data-latex` works also as `latex`
:::

::: {.fact data-latex='[Special]'}
Using beamer own definintion environment not in **bookdown** works with the same syntax
:::

Also this would work using `rmarkdown::beamer_presentation` without change
````

## Changes 

- [x] do not add \newtheorem in preamble when output is beamer format. This fixes the use of bookdown special environment that are using block2 knitr engine.

- [x] Only support specific format - _beamer_ defines its own environment so we don't let **bookdown** feature interfere.

- [x] Improve lua filter base on `latex-divs.lua` in **rmarkdown**

### CODE REVIEW

**Comment by XiangyunHuang** [comment]
all works well in **beamer**,  but not **book** documentclass.  

### Here is a minimal example.

````md
---
title: "book"
output: 
  bookdown::pdf_book:
    toc: no
    latex_engine: xelatex
    citation_package: natbib
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from book


::: {.theorem data-latex="[Metropolis]"}
[WARNING] data-latex attribute can't be used with one of bookdown custom environment. It has been removed.
:::


::: {.theorem name="Metropolis"}
NO WARNING. theorem is ok.
:::


## Some other blocks

::: {.lemma data-latex="[Metropolis]"}
[WARNING] data-latex attribute can't be used with one of bookdown custom environment. It has been removed.
:::


````


The output is ok.

截屏2021-04-21 下午7 31 58




### My session info 

```
xfun::session_info(c('bookdown', 'rmarkdown'))
R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16, RStudio 1.3.1093

Locale: en_US.UTF-8 / en_US.UTF-8 / en_US.UTF-8 / C / en_US.UTF-8 / en_US.UTF-8

Package version:
  base64enc_0.1.3   bookdown_0.21.11  digest_0.6.27     evaluate_0.14     glue_1.4.2       
  graphics_4.0.3    grDevices_4.0.3   highr_0.9         htmltools_0.5.1.1 jsonlite_1.7.2   
  knitr_1.32        magrittr_2.0.1    markdown_1.1      methods_4.0.3     mime_0.10        
  rlang_0.4.10      rmarkdown_2.7.12  stats_4.0.3       stringi_1.5.3     stringr_1.4.0    
  tinytex_0.31.4    tools_4.0.3       utils_4.0.3       xfun_0.22         yaml_2.2.1       

Pandoc version: 2.13
```

**Comment by cderv** [comment]
With `bookdown::pdf_book`, the syntax is the current working one 

````markdown
::: {.env #id name="NAME"}
:::
````

with 
 * `env` one of [bookdown supported environment](https://bookdown.org/yihui/bookdown/markdown-extensions-by-bookdown.html#theorems), ``
 * `#id` the identifier to use to reference theorem-like block
 * `name=` the name to append in parenthesis after the env name 

So you example should be 
````markdown
---
title: "book"
output: 
  bookdown::pdf_book:
    toc: no
    latex_engine: xelatex
    citation_package: natbib
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from book

::: {.theorem name="Metropolis"}
A theorem
:::


## Some other blocks

::: {.lemma name="Metropolis"}
A lemma
:::
```` 

To be clear, this is the way it is supposed to work without this PR. it works this way for HTML and PDF format. 
_beamer_ is not supported though because of its specificity - it is your issue. 

Do you expect the same syntax with `bookdown::beamer_presentation2` ?

**Comment by XiangyunHuang** [comment]
> Do you expect the same syntax with `bookdown::beamer_presentation2` ?

Yes. I feel better if they stay the same.

**Comment by cderv** [comment]
Ok I could change that. 

However, here are the limitations this will lead to: 

* bookdown's special syntax with `name=` attributes could be used with `beamer_presentation2` only for environment defined in _beamer_ and known by _bookdown_, as listed in https://github.com/rstudio/bookdown/issues/1143#issuecomment-823287646 (so theorem, lemma, corollary, definition, example, solution). 
* Other environment known by bookdown could be used **if** user defines it in preamble.tex (so proposition, conjecture, exercise, hypothesis, remark)
* For environment known only by _beamer_, the `data-latex=` or `latex=` syntax should be used. (so I think fact, problem, block and exampleblock). 

So 
* For common environment known by bookdown, the syntax would be the same between `pdf_document2`, `pdf_book` and `beamer_document2`
* For special _beamer_ env, the syntax would be different. Also `pdf_*` format would not about those. 

Is that ok ?

**Comment by XiangyunHuang** [comment]
OK, That's awesome! Thanks for your patient and great work!

File: R/latex.R
Comment (yihui) [inline R/latex.R:239]: Usually I don't use `isTRUE()` or `isFALSE()` on `getOption()`. If users set a strange option value that is not `TRUE` or `FALSE`, I'll just let R signal the error naturally so users know that they made a mistake.

```suggestion
  if (getOption("bookdown.theorem.preamble", TRUE) &&
```

BTW, did you mean `!` before `getOption()`? It feels like we want to skip adding the preamble when `options(bookdown.theorem.preamble = FALSE)`.

**Review by yihui — APPROVED** [review]
I didn't review the Lua script carefully. Feel free to merge whenever both you and @XiangyunHuang are happy with the PR. Thanks!

File: R/latex.R
Comment (cderv) [reply R/latex.R:239]: > If users set a strange option value that is not TRUE or FALSE, I'll just let R signal the error naturally so users know that they made a mistake.

Oh ok. I'll try to remember.

And yes we want to skip adding it when it is FALSE. thanks for catching that!

**Comment by cderv** [comment]
@XiangyunHuang let's try that then. I'll push the change for you to test.

@yihui this is only a matter of choice with the syntax to use for beamer and related potential limitation if we support the bookdown syntax. If you have an opinion feel free to share. Otherwise I'll merge it as soon as we are ok with @XiangyunHuang 

About the Lua filter, the changes are mainly to update based on `latex-div.lua` in **rmarkdown** as this is almost the same here that we are doing. It is also only about applying the filter for some known format to avoid issue with other unsupported format.

**Comment by cderv** [comment]
@XiangyunHuang if you reinstall new this PR this is the syntax to use 

````markdown
---
title: "book"
output: 
  bookdown::beamer_presentation2:
    toc: no
    latex_engine: xelatex
    theme: Xiaoshan
    citation_package: natbib
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from book

::: {.theorem name="Metropolis"}
A theorem
:::

::: {.lemma name="Metropolis"}
A lemma
:::

## Some other blocks

::: {.fact latex="[A fact]"}
Some interesting stuff
:::

::: {.problem latex=true}
A problem
:::
````

* `theorem` and `lemma` are supported by **bookdown**, so you can use the `name=` syntax
* `fact` and `problem` are only know by _beamer_ so you can use them in a beamer format but you need to use the classic custom block syntax from **rmarkdown** (`latex=` or `data-latex=`)

Are we merging this ? 😄

**Comment by XiangyunHuang** [comment]
@cderv  only one small problem, *theorem* and *lemma* block have extra blank line.

截屏2021-04-22 上午8 49 28

**Comment by cderv** [comment]
Yes, there is an extra line where add the `\label` part to allow referencing. That is the main value of **bookdown** feature. It seems this happens when a `theme` is used

I don't get this extraline when using default
image


Basically, the Lua filter here will insert a `\label` for referencing and it seems to cause this new lines when a beamer theme is used. This label is need in other format for the referencing. If I compile a document with this syntax in pure LaTeX, then we get the new line 

````markdown
---
title: "book"
output: 
  beamer_presentation:
    keep_tex: true
    toc: no
    latex_engine: xelatex
    theme: Xiaoshan
    citation_package: natbib
    template: null
link-citations: yes
colorlinks: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## some blocks from book

\begin{theorem}[Metropolis]
\label{thm:th1}A theorem with a label to reference
\end{theorem}

\begin{theorem}[Metropolis]
Another one without 
\end{theorem}

## Some other blocks

See Theorem \ref{thm:th1}
````
image

So what is the expected syntax in LaTeX for this ? 
Are referencing not used with _beamer_ theorem environment ? 

If they aren't I don't think there is any value of using **bookdown** special handling of this environment and that we should only allow the other syntax for beamer presentation. 

I could maybe remove this label insertion if there is no id provided in the div (no `#id`) but it will still insert a blank line if used with beamer.

**Comment by cderv** [comment]
> I could maybe remove this label insertion if there is no id provided in the div (no #id) but it will still insert a blank line if used with beamer.

I did that because we don't need to insert `\label{}` where there is no explicit id provided in the fenced div. 

This should work better for you now I guess. But there would still be the new line inserted if you provided an id to use as a label reference. 

Is this last change better ? 

@yihui I added manual test file for visual inspection that everything is ok. I hope it is fine.

**Comment by yihui** [comment]
> I added manual test file for visual inspection that everything is ok. I hope it is fine.

That's totally fine. Thanks!

**Comment by cderv** [comment]
@XiangyunHuang just waiting for you confirmation before merging this PR. Hope it is ok for you now. I think so but I prefer to check. Thanks !

**Comment by XiangyunHuang** [comment]
> @XiangyunHuang just waiting for you confirmation before merging this PR. Hope it is ok for you now. I think so but I prefer to check. Thanks !

It's OK. Thanks

**Pull request merged**
